storage/innobase/row/row0ins.cc

```diff
@@ -3391,63 +3409,134 @@ row_ins_step(
 inline
 void set_row_field_8(dtuple_t* row, int field_num, ib_uint64_t data, mem_heap_t* heap)
 {
+	static const ulint fsize = 8;
 	dfield_t* dfield = dtuple_get_nth_field(row, field_num);
-	byte* buf = static_cast<byte*>(mem_heap_alloc(heap, 8));
+	ut_ad(dfield->type.len == fsize);
+	byte* buf = static_cast<byte*>(mem_heap_alloc(heap, fsize));
 	mach_write_to_8(buf, data);
-	dfield_set_data(dfield, buf, 8);
+	dfield_set_data(dfield, buf, fsize);
 }
 
-#include "my_time.h"
-#include "sql_time.h"
-
 /***********************************************************//**
-Inserts a row to SYS_VTQ table.
-@return	error state */
-UNIV_INTERN
+Inserts a row to SYS_VTQ, low level.
+@return DB_SUCCESS if operation successfully completed, else error
+code */
+static __attribute__((nonnull, warn_unused_result))
 dberr_t
-vers_notify_vtq(que_thr_t* thr, mem_heap_t* heap)
+vers_row_ins_vtq_low(trx_t* trx, mem_heap_t* heap, dtuple_t* row)
 {
 	dberr_t		err;
-	trx_t*		trx = thr_get_trx(thr);
-	dict_table_t*	sys_vtq = dict_sys->sys_vtq;
-	ins_node_t*	node = ins_node_create(INS_DIRECT, sys_vtq, heap);
+	dtuple_t*	entry;
+	ulint		n_index = 0;
+	dict_index_t*	index	= dict_table_get_first_index(dict_sys->sys_vtq);
+	static const ulint flags
+		= (BTR_KEEP_SYS_FLAG
+			| BTR_NO_LOCKING_FLAG
+			| BTR_NO_UNDO_LOG_FLAG);
 
-	node->select = NULL;
-	node->values_list = NULL; // for INS_VALUES
+	entry = row_build_index_entry(row, NULL, index, heap);
+
+	dfield_t* dfield = dtuple_get_nth_field(entry, DATA_TRX_ID);
+	ut_ad(dfield->type.len == DATA_TRX_ID_LEN);
+	dfield_set_data(dfield, mem_heap_alloc(heap, DATA_TRX_ID_LEN), DATA_TRX_ID_LEN);
+	row_upd_index_entry_sys_field(entry, index, DATA_TRX_ID, trx->id);
+
+	err = row_ins_clust_index_entry_low(
+		flags, BTR_MODIFY_TREE, index, index->n_uniq, entry, 0, NULL, trx);
+
+	switch (err) {
+	case DB_SUCCESS:
+		break;
+	case DB_SUCCESS_LOCKED_REC:
+		/* The row had already been copied to the table. */
+		fprintf(stderr, "InnoDB: duplicate VTQ record!\n");
+		return DB_SUCCESS;
+	default:
+		return err;
+	}
+
+	mem_heap_t* offsets_heap = mem_heap_create(1024);
+
+	do {
```

Comment:
Late review.

We know exact VTQ table schema. And we know it has exactly 2 secondary index. Thus this loop should iterate exactly 2 times. And we expect DB_SUCCESS, otherwise it's fine to die in assert.